### PR TITLE
Kaetzchen keyserver

### PR DESCRIPTION
**Pull request by meskio**

### CODE REVIEW

**Comment by meskio** [comment]
I'm wondering if it will be nicer to do the key request as part of the `Send` function if needed, or is better that the clients do that more explicitly (like the API I'm proposing here).

I'm also wondering how useful is to have `GetKey` blocking until the response from the keyserver comes, as it might take several seconds (maybe more). We could figure out better ways to give feedback to the client if the request has failed.

File: java/client.go [hunk 1]
```diff
@@ -135,17 +140,63 @@ func (c Client) getMsg() (*Message, error) {
...
+					request.ch <- event.Err
+					continue
+				}
+
+				_, key, err := c.proxy.ParseKeyQueryResponse(event.Payload)
+				if err != nil {
+					request.ch <- err
+				} else {
+					request.ch <- c.proxy.SetRecipient(request.recipient, key)
+				}
```
Comment (mixmasala) [inline java/client.go]: at this point the map entry can be deleted

File: java/client.go [hunk 2]
```diff
@@ -135,17 +140,63 @@ func (c Client) getMsg() (*Message, error) {
...
 
+// GetKey for recipient and add it to the local key storage
+func (c Client) GetKey(recipient string) error {
+	msgID, err := c.proxy.QueryKeyFromProvider(c.address, recipient)
+	if err != nil {
+		return err
+	}
+	ch := make(chan error)
+	c.kaetzchenCh <- kaetzchenKeyRequest{recipient, msgID, ch}
+	return <-ch
```
Comment (mixmasala) [inline java/client.go]: there should probably be a timeout in case the KaetzchenKeyRequest or reply is lost.

**Comment by david415** [comment]
we instead want to add this feature of automatically retrieving keys into the mailproxy api.

**Pull request closed without merging**